Source/DafnyCore/Compilers/Rust/Dafny-compiler-rust.dfy

```diff
@@ -150,38 +555,93 @@ module {:extern "DCOMP"} DCOMP {
             case Path(traitPath, typeArgs, Trait(_)) => {
               var pathStr := GenPath(traitPath);
               var typeArgs := GenTypeArgs(typeArgs, false, false);
-              var body := "";
+              var body: seq<R.ImplMember> := [];
               if traitPath in traitBodies {
                 body := traitBodies[traitPath];
               }
 
               var genSelfPath := GenPath(path);
-              s := s + "\nimpl " + constrainedTypeParams + pathStr + typeArgs + " for ::std::rc::Rc<" + genSelfPath + typeParams + ">" + whereConstraints + " {\n" + body + "\n}";
+              var x := R.ImplDecl(
+                R.ImplFor(
+                  sConstrainedTypeParams,
+                  R.RawType(pathStr + typeArgs),
+                  R.RawType("::std::rc::Rc<" + genSelfPath + typeParams + ">"),
+                  whereConstraints,
+                  body
+                ));
+              s := s + [x];
             }
             case _ => {}
           }
           i := i + 1;
         }
       }
 
-      var defaultImpl := "impl " + constrainedTypeParams + " ::std::default::Default for " + escapeIdent(c.name) + typeParams + whereConstraints + " {\n";
-      defaultImpl := defaultImpl + "fn default() -> Self {\n";
-      defaultImpl := defaultImpl + escapeIdent(c.name) + "::new()\n";
-      defaultImpl := defaultImpl + "}\n";
-      defaultImpl := defaultImpl + "}\n";
-
-      var printImpl := "impl " + constrainedTypeParams + " ::dafny_runtime::DafnyPrint for " + escapeIdent(c.name) + typeParams + " {\n" + "fn fmt_print(&self, __fmt_print_formatter: &mut ::std::fmt::Formatter, _in_seq: bool) -> std::fmt::Result {\n";
-      printImpl := printImpl + "write!(__fmt_print_formatter, \"" + c.enclosingModule.id + "." + c.name + "\")\n}\n}\n";
-
-      var ptrPartialEqImpl := "impl " + typeParams + " ::std::cmp::PartialEq for " + escapeIdent(c.name) + typeParams + " {\n";
-      ptrPartialEqImpl := ptrPartialEqImpl + "fn eq(&self, other: &Self) -> bool {\n";
-      ptrPartialEqImpl := ptrPartialEqImpl + "::std::ptr::eq(self, other)";
-      ptrPartialEqImpl := ptrPartialEqImpl + "\n}\n}\n";
-
-      var identEraseImpls := "impl " + constrainedTypeParams + " ::dafny_runtime::DafnyErasable for " + escapeIdent(c.name) + typeParams + " {\n" + "type Erased = " + escapeIdent(c.name) + erasedParams + ";\n}\n";
-      identEraseImpls := identEraseImpls + "impl " + constrainedEraseParams + " ::dafny_runtime::DafnyUnerasable<" + escapeIdent(c.name) + unerasedParams + "> for " + escapeIdent(c.name) + typeParams + " {}\n";
-
-      s := s + "\n" + defaultImpl + "\n" + printImpl + "\n" + ptrPartialEqImpl + "\n" + identEraseImpls;
+      var d := R.ImplFor(
```

Comment:
Consider occasionally adding high-level comments to show the rough template for what we're generating, as we do in the C# implementation of rewriters and code generators. The string manipulation version could have already benefited from that, and I think the benefit is even greater when you start manipulating AST nodes instead.

Reply:
I might not do this immediately as I'm constantly evolving the representation, but please remind me of doing that when the implementation is more stabilized if I have not done so yet.